### PR TITLE
AT-21 Emoji flow

### PR DESCRIPTION
**Pull request by RamsesDijkstra**
This is a first demo of the emoji flow with the megamoji disabled since it is still a WIP. This PR is made so Kamen can test this on dev together with Mitia his frontend work.

### CODE REVIEW

File: internal/storage/db.go [hunk 1]
```diff
@@ -23,6 +23,7 @@ const (
 	getDefaultEntranceWorldQuery                    = `SELECT value FROM node_settings WHERE name = 'EntranceWorld';`
 	getWorldByURLQuery                              = `SELECT worldId FROM url_mapping WHERE URL = ?;`
 	getParentWorldQuery                             = `SELECT GetParentWorldByID(id) FROM spaces WHERE id = ?;`
+	getOnlineUserByUserID                           = `SELECT spaceId FROM online_users WHERE userId = ?;`
```
Comment (cnaize) [inline internal/storage/db.go:26]: is it used somewhere?

File: internal/universe/interactions.go [hunk 2]
```diff
@@ -2,12 +2,38 @@ package universe
...
+	TimeStamp    time.Time
+	Count        uint32
+	PondPosition cmath.Vec3
+	UserPosition cmath.Vec3
+	EmojiID      string
+}
+
+var ponds []Pond
+var triggerInterval time.Time = time.Now().Add(time.Second * 10)
+var triggerCount uint32 = 10
```
Comment (cnaize) [inline internal/universe/interactions.go:35]: why it's here? is it supposed to be on database layer?

File: internal/universe/interactions.go [hunk 3]
```diff
@@ -2,12 +2,38 @@ package universe
...
+	EmojiID    string `json:"emojiID"`
+}
+
+type Pond struct {
+	TimeStamp    time.Time
+	Count        uint32
+	PondPosition cmath.Vec3
+	UserPosition cmath.Vec3
+	EmojiID      string
+}
```
Comment (cnaize) [inline internal/universe/interactions.go:31]: why it's here? is it supposed to be on database layer?

File: internal/universe/interactions.go (hunk 2 above)
Comment (cnaize) [inline internal/universe/interactions.go:35]: please get rid of global vars

File: internal/universe/interactions.go [hunk 4]
```diff
@@ -64,6 +90,73 @@ func (u *User) InteractionHandler(m *posbus.TriggerInteraction) {
 	}
 }
 
+func (u *User) HandleEmoji(m *posbus.RelayToController) {
```
Comment (cnaize) [inline internal/universe/interactions.go:93]: why this method doesn't returns error?

File: internal/universe/interactions.go [hunk 5]
```diff
@@ -64,6 +90,73 @@ func (u *User) InteractionHandler(m *posbus.TriggerInteraction) {
...
+	msgData := m.Data()
+	log.Info("Got msg data: ", msgData)
+
+	var emojiMessage EmojiMessage
+	err := json.Unmarshal(msgData, &emojiMessage)
+	if err != nil {
+		return
```
Comment (cnaize) [inline internal/universe/interactions.go:100]: please return error with description here

File: internal/universe/interactions.go [hunk 6]
```diff
@@ -64,6 +90,73 @@ func (u *User) InteractionHandler(m *posbus.TriggerInteraction) {
...
+				"emoji", msgData,
+			).WebsocketMessage(),
+		)
+	} else {
+		u.world.Broadcast(
+			posbus.NewRelayToUnityMsg(
+				"emoji", msgData,
+			).WebsocketMessage(),
+		)
+		s, _ := u.world.spaces.GetPresent(spaceID)
```
Comment (cnaize) [inline internal/universe/interactions.go:118]: error swallowed

File: internal/universe/interactions.go (hunk 2 above)
Comment (RamsesDijkstra) [reply internal/universe/interactions.go:35]: done

File: internal/universe/interactions.go (hunk 3 above)
Comment (RamsesDijkstra) [reply internal/universe/interactions.go:31]: I removed line 13-15 for now. EmojiMessage is used to Unmarshal the Json that comes in from Unity. This has nothing to do with DB layer. Pond is a stub that will be used with megamoji logic once i can continue to work on that.

File: internal/storage/db.go (hunk 1 above)
Comment (RamsesDijkstra) [reply internal/storage/db.go:26]: removed

File: internal/universe/interactions.go (hunk 4 above)
Comment (RamsesDijkstra) [reply internal/universe/interactions.go:93]: added error return type with error message

File: internal/universe/interactions.go (hunk 5 above)
Comment (RamsesDijkstra) [reply internal/universe/interactions.go:100]: added description to error

File: internal/universe/interactions.go (hunk 6 above)
Comment (RamsesDijkstra) [reply internal/universe/interactions.go:118]: i checked other functions that implement GetPresent() and handled the error the same way.

File: internal/universe/interactions.go [hunk 7]
```diff
@@ -64,6 +81,82 @@ func (u *User) InteractionHandler(m *posbus.TriggerInteraction) {
...
 
+func (u *User) HandleEmoji(m *posbus.RelayToController) error {
+	msgData := m.Data()
+	log.Info("Got msg data: ", msgData)
+
+	var emojiMessage EmojiMessage
+	err := json.Unmarshal(msgData, &emojiMessage)
+	if err != nil {
+		return errors.WithMessage(err, "failed to unmarshal emoji payload")
+	}
```
Comment (cnaize) [inline internal/universe/interactions.go:92]: put it in oneliner please

File: internal/universe/interactions.go [hunk 8]
```diff
@@ -64,6 +81,82 @@ func (u *User) InteractionHandler(m *posbus.TriggerInteraction) {
...
+		)
+	} else {
+		u.world.Broadcast(
+			posbus.NewRelayToUnityMsg(
+				"emoji", msgData,
+			).WebsocketMessage(),
+		)
+		s, ok := u.world.spaces.GetPresent(spaceID)
+		if !ok {
+			log.Info("failed to get present space ", spaceID.String())
```
Comment (cnaize) [inline internal/universe/interactions.go:111]: it should be error or warning at least
and please use errros.WithMessage for describing errors

File: internal/universe/interactions.go (hunk 7 above)
Comment (RamsesDijkstra) [reply internal/universe/interactions.go:92]: done

File: internal/universe/interactions.go [hunk 9]
```diff
@@ -86,6 +103,82 @@ func (u *User) InteractionHandler(m *posbus.TriggerInteraction) {
...
+		if p.EmojiID == emojiMessage.EmojiID {
+			pondExists = true
+			// existing pond
+			currentPond = p
+		}
+	}
+
+	if !pondExists {
+		// new pond
+		currentPond = Pond{time.Now(), 0, cmath.Vec3{0, 0, 0}, cmath.Vec3{0, 0, 0}, emojiMessage.EmojiID}
```
Comment (cnaize) [inline internal/universe/interactions.go:164]: it's better to use predefined funcs like MNan32Vec3

**Pull request merged**